File: leetcode/0725-split-linked-list-in-parts/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int length = 0;
        for (ListNode node = head; node != null; node = node.next) {
            length++;
        }

        int baseSize = length / k;
        int extra = length % k;

        ListNode[] result = new ListNode[k];
        ListNode curr = head;
        for (int i = 0; i < k; i++) {
            int partSize = baseSize + (i < extra ? 1 : 0);
            if (partSize == 0) {
                result[i] = null;
                continue;
            }

            ListNode partHead = curr;
            for (int j = 0; j < partSize - 1; j++) {
                curr = curr.next;
            }
            ListNode nextPart = curr.next;
            curr.next = null;
            curr = nextPart;
            result[i] = partHead;
        }

        return result;
    }
}

File: leetcode/0725-split-linked-list-in-parts/solution.py
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        length = 0
        node = head
        while node:
            length += 1
            node = node.next

        base_size, extra = divmod(length, k)

        result = []
        curr = head
        for i in range(k):
            part_size = base_size + (1 if i < extra else 0)
            if part_size == 0:
                result.append(None)
                continue

            part_head = curr
            for _ in range(part_size - 1):
                curr = curr.next
            next_part = curr.next
            curr.next = None
            curr = next_part
            result.append(part_head)

        return result
